### ISSUE TITLE
didCreateToken method not invoked

### ISSUE THREAD

**Issue by greggmojica** [stripe/stripe-ios#1400]
I'm building a simple app that needs to use stripe for payments. In my viewDidLoad() method I have the following code:

```
let addCardViewController = STPAddCardViewController()
addCardViewController.delegate = self
navigationController?.pushViewController(addCardViewController, animated: true)
```

I have the following delegate methods:

```
extension Checkout: STPAddCardViewControllerDelegate {
    func addCardViewControllerDidCancel(_ addCardViewController: STPAddCardViewController) {
        navigationController?.popViewController(animated: true)
    }

    private func addCardViewController(_ addCardViewController: STPAddCardViewController, didCreateToken token: STPToken, completion: @escaping STPErrorBlock) {
        StripeClient.shared.completeCharge(with: token, amount: 50) { result in
            switch result {
                case .success:
                    completion(nil)
                    let alertController = UIAlertController(title: "Congrats", message: "Your payment was successful!", preferredStyle: .alert)
                    let alertAction = UIAlertAction(title: "OK", style: .default, handler: { _ in
                      self.navigationController?.popViewController(animated: true)
                    })
                    alertController.addAction(alertAction)
                    self.present(alertController, animated: true)
            case .failure(let error):
                completion(error)
            }
        }
    }
}
```
However, the didCreateToken method is never invoked and I'm not sure why. Help would be appreciated!

**Comment by yuki-stripe** [comment]
Hi @greggmojica,

Could you share the version of the SDK you're using?  `didCreateToken` was deprecated in v16

**Comment by greggmojica** [comment]
@yuki-stripe I'm using the latest version from cocoapods. If this method is deprecated, how can one generate a token using the STPAddCardViewController currently?

**Comment by yuki-stripe** [comment]
@greggmojica It's no longer possible - since version 16, `STPAddCardViewController` and other UI components migrated from Tokens/Sources to [PaymentMethods](https://stripe.com/docs/payments/payment-methods#transitioning).  I recommend you [migrate to the Payment Methods + Payment Intents API](https://stripe.com/docs/payments/payment-intents/migration/automatic-confirmation).  This is especially important [if you're affected by SCA](https://stripe.com/docs/strong-customer-authentication/doineed).

If you can't migrate to PaymentMethods, I recommend you use a version prior to v16.

**Comment by Igorsnaki** [comment]
@yuki-stripe Hi , I also have the same issue. I used 15.0.1 ios Stripe sdk and use method - 

(void)addCardViewController:(STPAddCardViewController *)addCardViewController
               didCreateToken:(STPToken *)token
                   completion:(STPErrorBlock)completion

I need **token.tokenId**

but starting from Xcode 11.4 - ios Stripe SDK doesn't compile. 

You fix this only in 19.0.1 

Could you please help me? 


- (void)addCardViewController:(STPAddCardViewController *)addCardViewController
       didCreatePaymentMethod:(STPPaymentMethod *)paymentMethod
                   completion:(STPErrorBlock)completion;

Can I somehow get **token.tokenId** ( which I get in  didCreateToken  method)  from  paymentMethod ???

**Comment by abdulmanan-whizpool** [comment]
// Pass it to STPAPIClient to create a Token
        STPAPIClient.shared().createToken(withCard: cardParams) { token, error in
            guard let token = token else {
                // Handle the error
                return
            }
            let tokenID = token.tokenId
            // Send the token identifier to your server...
            print(token)
        } 
this method returns token 
but you have to add card values Manually
